plonk/examples/proof_of_exp.rs

```diff
@@ -0,0 +1,132 @@
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! This file contains an example showing how to build a proof of knowledge
+//! of the exponent over a native field.
+//!
+//! - secret input `x`;
+//! - public generator `G`;
+//! - public group element `X := xG`
+
+use ark_bls12_381::Bls12_381;
+use ark_ec::{
+    twisted_edwards_extended::GroupAffine as TEAffine, AffineCurve, ModelParameters,
+    ProjectiveCurve, TEModelParameters,
+};
+use ark_ed_on_bls12_381::{EdwardsAffine, Fr};
+use ark_ff::PrimeField;
+use ark_std::{test_rng, UniformRand};
+use jf_plonk::{
+    circuit::{customized::ecc::Point, Circuit, PlonkCircuit},
+    errors::PlonkError,
+    proof_system::{PlonkKzgSnark, Snark},
+    transcript::StandardTranscript,
+};
+use jf_utils::fr_to_fq;
+
+// The following example proves knowledge of exponent.
+#[allow(non_snake_case)]
+fn main() -> Result<(), PlonkError> {
+    // set up the inputs and parameters
+    let mut rng = test_rng();
+    let x = Fr::rand(&mut rng);
+    let G = EdwardsAffine::prime_subgroup_generator();
+    let X = G.mul(x).into_affine();
+
+    // Our first step is to build a circuit for the following statements.
+    // - secret input `x`;
+    // - public generator `G`;
+    // - public group element `X := xG`
+    // This circuit does not need to have real inputs.
+    // We can simply use a dummy data set.
+    let circuit = proof_of_exponent_circuit(x, X)?;
+
+    // Knowing the circuit size, we are able to simulate the universal
+    // setup and obtain the structured reference string (SRS).
+    //
+    let circuit_size = circuit.num_gates() + 100; // adding 100 as a buffer
+    let srs = PlonkKzgSnark::<Bls12_381>::universal_setup(circuit_size, &mut rng)?;
+
+    // Then, we generate the proving key and verification key from the SRS and
+    // circuit.
+    let (pk, vk) = PlonkKzgSnark::<Bls12_381>::preprocess(&srs, &circuit)?;
+
+    // Next, we generate the proof.
+    // The proof generation will need an internal transcript for Fiat-Shamir
+    // transformation. For this example we use a `StandardTranscript`.
+    let proof = PlonkKzgSnark::<Bls12_381>::prove::<_, _, StandardTranscript>(
+        &mut rng, &circuit, &pk, None,
+    )?;
+
+    // Last step, verify the proof against the public inputs.
+    assert!(PlonkKzgSnark::<Bls12_381>::verify::<StandardTranscript>(
+        &vk,
+        &[X.x, X.y],
+        &proof,
+        None,
+    )
+    .is_ok());
+
+    Ok(())
+}
+
+// This function build the PoE circuit.
+//
+// We write the code with generics so that is can be adapted to
+// multiple curves.
+// Specifically, the PoE is associated with an embedded curve with param `P`,
+// that defined twisted-edwards parameters for a curve.
+#[allow(non_snake_case)]
+fn proof_of_exponent_circuit<P>(
+    x: P::ScalarField,
+    X: TEAffine<P>,
+) -> Result<PlonkCircuit<P::BaseField>, PlonkError>
```

Comment:
can I suggest we change this function to be generic over two parameters (one for twisted edward curve, one for Bls12-381), reason being people's default mindset would be getting a circuit over a scalar field (but here we return a `PlonkCircuit<P::BaseField>`, not immediately clear that the `P` here is for the `TEParam`. 

```rust
fn proof_of_exponent_circuit<P, P2>(
    x: P::ScalarField,
    X: TEAffine<P>,
) -> Result<PlonkCircuit<P2::ScalarField>, PlonkError>
where
    P: TEModelParameters + Clone,
    P2: SWModelParameters<ScalarField = P::BaseField>,
    <P as ModelParameters>::BaseField: PrimeField,
```

of course, renaming `<P, P2>` to `<EmbedParam, CurveParam>` or better name is definitely good.

Reply:
sounds good.

Reply:
So here is the new one
``` rust
// Specifically, the PoE is associated with
// - an embedded curve with param `EmbedCurve` that defined twisted-edwards
//   parameters for a curve
// - a pairing engine
// - the native field F for the prove system
#[allow(non_snake_case)]
fn proof_of_exponent_circuit<EmbedCurve, PairingCurve, F>
```

Reply:
is the `F` necessary?

Reply:
F is to ensure that this is native field evaluation. It is not really `necessary` the same way as `PairingCurve`, but is nice to have IMHO. WDYT? 

``` rust
where
    F: PrimeField,
    EmbedCurve: TEModelParameters<BaseField = F> + Clone,
    PairingCurve: PairingEngine<Fr = F>,
```

Reply:
Okay, sure, make sense. 

(my only minor concern is that my suggestion is making this function looks more complicated than it was, if we were to have a non-generic example, it would looks much simpler which might be a plus for example code)